tripper/literal.py

```diff
@@ -0,0 +1,193 @@
+        elif self.datatype and self.datatype not in (
+            RDF.PlainLiteral,
+            RDF.XMLLiteral,
+            RDFS.Literal,
+            XSD.anyURI,
+            XSD.language,
+            XSD.Name,
+            XSD.NMName,
+            XSD.normalizedString,
+            XSD.string,
+            XSD.token,
+            XSD.NMTOKEN,
+        ):
+            warnings.warn(f"unknown datatype: {self.datatype} - assuming string")
+        return value
+
+    def n3(self):  # pylint: disable=invalid-name
+        """Returns a representation in n3 format."""
+        if self.lang:
+            return f'"{self}"@{self.lang}'
+        if self.datatype:
+            return f'"{self}"^^{self.datatype}'
+        return f'"{self}"'
+
+
+def en(value):  # pylint: disable=invalid-name
+    """Convenience function that returns value as a plain english literal.
+
+    Equivalent to``Literal(value, lang="en")``.
+    """
+    return Literal(value, lang="en")
+
+
+def parse_literal(literal: "Any") -> "Literal":
+    """Parse `literal` and return it as an instance of Literal.
+
+    The main difference between this function and the Literal constructor,
+    is that this function correctly interprets n3-encoded literal strings.
+    """
+    # pylint: disable=invalid-name,too-many-branches
+    lang, datatype = None, None
+
+    if isinstance(literal, Literal):
+        return literal
+
+    if not isinstance(literal, str):
+        for type_, datatype in Literal.datatypes.items():
+            if isinstance(literal, type_):
+                return Literal(literal, lang=lang, datatype=datatype)
+        TypeError(f"unsupported literal type: {type(literal)}")
+
+    match = re.match(r'^\s*("""(.*)"""|"(.*)")\s*$', literal, flags=re.DOTALL)
+    if match:
+        _, v1, v2 = match.groups()
+        value, datatype = v1 if v1 else v2, XSD.string
+    else:
+        match = re.match(
+            r'^\s*("""(.*)"""|"(.*)")\^\^(.*)\s*$', literal, flags=re.DOTALL
+        )
+        if match:
+            _, v1, v2, datatype = match.groups()
+            value = v1 if v1 else v2
+        else:
+            match = re.match(
+                r'^\s*("""(.*)"""|"(.*)")@(.*)\s*$', literal, flags=re.DOTALL
+            )
+            if match:
+                _, v1, v2, lang = match.groups()
+                value = v1 if v1 else v2
+            else:
+                value = literal
+
+    if lang or datatype:
+        if datatype:
+            types = {v: k for k, v in Literal.datatypes.items()}
+            type_ = types[datatype]
+            try:
+                value = type_(value)
+            except TypeError:
+                pass
+        return Literal(value, lang=lang, datatype=datatype)
+
+    for type_, datatype in Literal.datatypes.items():
+        if type_ is not bool:
+            try:
+                return Literal(type_(literal), lang=lang, datatype=datatype)
+            except (ValueError, TypeError):
+                pass
```

Comment:
Why do we allow these exceptions to go uncaught? Wouldn't that constitute not being able to parse the `literal` and we should therefore raise?

Reply:
The literal does not provide datatype, so we try to infer the datatype by looping over all supported datatypes. If we get a ValueError or TypeError it is not the correct datatype and we try again...